### PR TITLE
pythonPackages.python-miio: fix build

### PR DESCRIPTION
**Pull request by georgewhewell**
###### Motivation for this change
python-miio does not build-

```
ERROR: No matching distribution found for pytz<2020.0,>=2019.3 (from python-miio==0.5.3)
```

###### Things done
patch `setup.py` so any version is allowed
add some new dependencies
remove failing (broken?) test

testing:
- remaining tests pass on python3.8
- exeutables seem to function correctly

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/python-miio/default.nix [hunk 1]
```diff
@@ -22,10 +24,20 @@ buildPythonPackage rec {
...
+    # use any available version from nixpkgs
+    substituteInPlace setup.py \
+      --replace "pytz>=2019.3,<2020.0" "pytz" \
+      --replace "cryptography>=2.9,<3.0" "cryptography" \
+      --replace "zeroconf>=0.25.1,<0.26.0" "zeroconf"
+  '';
+
   checkInputs = [ pytest ];
-  propagatedBuildInputs = [ appdirs click construct cryptography zeroconf attrs pytz tqdm netifaces ];
+  propagatedBuildInputs = [ appdirs click construct cryptography zeroconf attrs pytz tqdm netifaces croniter importlib-metadata ];
```
Comment (jonringer) [inline pkgs/development/python-modules/python-miio/default.nix:36]: importlib-metadata is only needed python<=3.7
```suggestion
  propagatedBuildInputs = [ appdirs click construct cryptography zeroconf attrs pytz tqdm netifaces croniter ]
    ++ lib.optionals (pythonOlder "3.8") [ importlib-metadata ];
```

File: pkgs/development/python-modules/python-miio/default.nix [hunk 2]
```diff
@@ -22,10 +24,20 @@ buildPythonPackage rec {
...
 
   checkPhase = ''
+    # tests/test_device.py appears to be broken
```
Comment (jonringer) [inline pkgs/development/python-modules/python-miio/default.nix:39]: comments within code blocks can cause rebuilds if changed, preference to have them as nix comments
```suggestion
  # tests/test_device.py appears to be broken
  checkPhase = ''
```

**Comment by georgewhewell** [comment]
@jonringer thanks, applied and rebased

File: pkgs/development/python-modules/python-miio/default.nix [hunk 3]
```diff
@@ -11,6 +11,8 @@
 , pytz
 , tqdm
 , netifaces
+, croniter
+, importlib-metadata
```
Comment (jonringer) [inline pkgs/development/python-modules/python-miio/default.nix:15]:
```suggestion
, importlib-metadata
, lib
```

```
error: while evaluating anonymous function at /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/lib/meta.nix:46:54, called from undefined position:
while evaluating 'isDerivation' at /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/lib/attrsets.nix:305:18, called from /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/lib/meta.nix:46:62:
while evaluating 'callPackageWith' at /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/lib/customisation.nix:117:35, called from /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/pkgs/top-level/python-packages.nix:5761:17:
while evaluating 'makeOverridable' at /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/lib/customisation.nix:67:24, called from /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/lib/customisation.nix:121:8:
undefined variable 'lib' at /home/jon/.cache/nixpkgs-review/pr-99585-1/nixpkgs/pkgs/development/python-modules/python-miio/default.nix:37:8
```

**Comment by georgewhewell** [comment]
@jonringer `importlib_metadata` seems to be required on python 3.8:

```
Traceback (most recent call last):
  File "/nix/store/2xaplvp6mhnmacxx1z0xcxrn361mbc36-python3.8-python-miio-0.5.3/bin/.mirobo-wrapped", line 6, in <module>
    from miio.vacuum_cli import cli
  File "/nix/store/2xaplvp6mhnmacxx1z0xcxrn361mbc36-python3.8-python-miio-0.5.3/lib/python3.8/site-packages/miio/__init__.py", line 2, in <module>
    from importlib_metadata import version  # type: ignore
ModuleNotFoundError: No module named 'importlib_metadata'
```

**Comment by jonringer** [comment]
then upstream incorrectly packaged it, it's part of the standard library, and should not be imported

**Comment by georgewhewell** [comment]
superseded by https://github.com/NixOS/nixpkgs/pull/102183

**Pull request closed without merging**